python/tests/test_topology.py

```diff
@@ -97,6 +98,618 @@ def generate_segments(n, sequence_length=100, seed=None):
+            for v in children:
+                stack.append((v, depth + 1, tree.time(v)))
+            path_distance[tree_index][u] = depth
+            time_distance[tree_index][u] = tree.time(tree.root) - time
+            if len(tree.children(u)) == 0:
+                M[tree_index][u + n] = tree.branch_length(u)
+
+        for index, n1 in enumerate(samples):
+            for n2 in samples[index + 1:]:
+                mrca = tree.mrca(n1, n2)
+                pair_index = n1 * (n1 - 2 * k + 1) // -2 + n2 - n1 - 1
+                assert m[tree_index][pair_index] == 1
+                m[tree_index][pair_index] = path_distance[tree_index][mrca]
+                M[tree_index][pair_index] = time_distance[tree_index][mrca]
+
+    vT1 = 0
+    vT2 = 0
+    distance_sum = 0
+
+    for i in range(n + k):
+        vT1 = (m[0][i] * (1 - lambda_param)) + (lambda_param * M[0][i])
+        vT2 = (m[1][i] * (1 - lambda_param)) + (lambda_param * M[1][i])
+        distance_sum += (vT1 - vT2) ** 2
+
+    return math.sqrt(distance_sum)
+
+
+class TestKCMetric(unittest.TestCase):
+    """
+    Tests on the KC metric distances.
+    """
+    def test_same_tree_zero_distance(self):
+        for n in range(2, 10):
+            for seed in range(1, 10):
+                ts = msprime.simulate(n, random_seed=seed)
+                tree = ts.first()
+                self.assertEqual(kc_distance_tree(tree, tree), 0)
+                self.assertEqual(kc_distance_tree_simple(tree, tree), 0)
+                ts = msprime.simulate(n, random_seed=seed)
+                tree2 = ts.first()
+                self.assertEqual(kc_distance_tree(tree, tree2), 0)
+                self.assertEqual(kc_distance_tree_simple(tree, tree2), 0)
+
+    def test_sample_2_zero_distance(self):
+        # All trees with 2 leaves must be equal distance from each other.
+        for seed in range(1, 10):
+            tree1 = msprime.simulate(2, random_seed=seed).first()
+            tree2 = msprime.simulate(2, random_seed=seed + 1).first()
+            self.assertEqual(kc_distance_tree(tree1, tree2), 0)
+            self.assertEqual(kc_distance_tree_simple(tree1, tree2), 0)
+
+    def test_different_samples_error(self):
+        tree1 = msprime.simulate(10, random_seed=1).first()
+        tree2 = msprime.simulate(2, random_seed=1).first()
+        self.assertRaises(ValueError, kc_distance_tree, tree1, tree2)
+        self.assertRaises(ValueError, kc_distance_tree_simple, tree1, tree2)
+
+    def validate_trees(self, n):
+        for seed in range(1, 10):
+            tree1 = msprime.simulate(n, random_seed=seed).first()
+            tree2 = msprime.simulate(n, random_seed=seed + 1).first()
+            self.assertAlmostEqual(
+                kc_distance_tree(tree1, tree2), kc_distance_tree(tree1, tree2))
+            self.assertAlmostEqual(
+                kc_distance_tree_simple(tree1, tree2),
+                kc_distance_tree_simple(tree1, tree2))
+
+    def test_sample_3(self):
```

Comment:
I use pytest parameterized tests which basically does this without the code duplication. however we're not using pytest. https://docs.pytest.org/en/latest/parametrize.html

Reply:
Pytest looks like all right, but it'd be a big job moving over now. There's not a lot of motivation for taking this work on.

Reply:
Here's another solution I came across which would fit well, and doesn't require any extra packages: https://docs.python.org/3/library/unittest.html#distinguishing-test-iterations-using-subtests.

From https://stackoverflow.com/questions/32899/how-do-you-generate-dynamic-parameterized-unit-tests-in-python

Reply:
Hey, this is great feature I didn't know about, thanks @brianzhang01! I'll definitely be deploying that one.